sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -382,7 +382,14 @@ object ColumnPruning extends Rule[LogicalPlan] {
       val required = child.references ++ p.references
       if ((child.inputSet -- required).nonEmpty) {
         val newChildren = child.children.map(c => prunedChild(c, required))
-        p.copy(child = child.withNewChildren(newChildren))
+        val newChild = child.withNewChildren(newChildren)
+        val sameOutput = newChild.output.size == p.output.size &&
+          newChild.output.zip(p.output).forall(pair => pair._1.semanticEquals(pair._2))
```

Comment:
yeah, they have the same effect. Do you think it is better to add this check in this pattern case?